definitions/npm/styled-components_v2.x.x/flow_v0.53.x-/styled-components_v2.x.x.js

```diff
@@ -0,0 +1,332 @@
+// @flow
+
+type $npm$styledComponents$Interpolation = ((executionContext: Object) => string) | string | number;
+type $npm$styledComponents$NameGenerator = (hash: number) => string;
+
+type $npm$styledComponents$TaggedTemplateLiteral<R> = {| (Array<string>, $npm$styledComponents$Interpolation): R |};
+
+type $npm$styledComponents$ReactComponentFunctional<Props: {}> = Props => React$Element<*>;
+type $npm$styledComponents$ReactComponentClass<Props, DefaultProps = *> = Class<React$Component<DefaultProps & Props, *>>
+
+type $npm$styledComponents$ReactComponentUnion<Props> =
```

Comment:
I tried to work a bit on it as well, but the code really seems a bit hard to understand so I felt rather uncomfortable. But it really seemed like it could be simplified a lot.

f.e. i rewrote the withTheme stuff like this (following the documentation on the flow site for injection of props), but as this replaces 15 lines with 3 I was afraid I was missing something.

declare function $npm$styledComponents$WithTheme<Props: {}>(
  Component: React$ComponentType<{ theme: $npm$styledComponents$Theme & Props}>
) : React$ComponentType<Props>